tools/test-proxy/Azure.Sdk.Tools.TestProxy/Store/GitStore.cs

```diff
@@ -2,36 +2,513 @@
+    /// </summary>
     public class GitStore : IAssetsStore
     {
-        public void Push(string pathToAssetsJson, string contextPath) {
-            var config = ParseConfigurationFile(pathToAssetsJson);
+        private HttpClient httpClient = new HttpClient();
+        public GitProcessHandler GitHandler = new GitProcessHandler();
+        public string DefaultBranch = "main";
+        public string FileName = "assets.json";
+
+        public GitStore() { }
+
+        public GitStore(GitProcessHandler processHandler) {
+            GitHandler = processHandler;
+        }
+
+        #region push, reset, restore, and other asset repo implementations 
+        /// <summary>
+        /// Pushes a set of changed files to the assets repo. Honors configuration of assets.json passed into it.
+        /// </summary>
+        /// <param name="pathToAssetsJson"></param>
+        /// <returns></returns>
+        public async Task Push(string pathToAssetsJson) {
+            var config = await ParseConfigurationFile(pathToAssetsJson);
+            var pendingChanges = DetectPendingChanges(config);
+            var onLatestSHA = true;
+
+            if (pendingChanges.Length > 0)
+            {
+                if(!GitHandler.TryRun(config, $"rev-parse origin/{config.AssetsRepoBranch}", out CommandResult result))
+                {
+                    // if we have a nonzero exit code, only code 128 is acceptable.
+                    if (result.ExitCode != 128)
+                    {
+                        throw GenerateInvokeException(result);
+                    }
+                }
+                else
+                {
+                    var retrievedSHA = result.StdOut.Trim();
+                    if (retrievedSHA != config.SHA)
+                    {
+                        onLatestSHA = false;
+                    }
+                }
+
+                if (onLatestSHA)
+                {
+                    if (!GitHandler.TryRun(config, $"checkout -b {config.AssetsRepoBranch}", out CommandResult checkoutResult))
+                    {
+                        throw GenerateInvokeException(checkoutResult);
+                    }
+
+                    if (!GitHandler.TryRun(config, $"add -A .", out CommandResult addResult))
+                    {
+                        throw GenerateInvokeException(addResult);
+                    }
+
+                    if (!GitHandler.TryRun(config, $"commit -m \"Automatic asset update from test-proxy.\"", out CommandResult commitResult))
+                    {
+                        throw GenerateInvokeException(commitResult);
+                    }
+
+                    if (!GitHandler.TryRun(config, $"push origin {config.AssetsRepoBranch}", out CommandResult pushResult))
+                    {
+                        throw GenerateInvokeException(pushResult);
+                    }
```

Comment:
It might be nice to have an alternate `Run` method which throws an exception so users of this code could write back to back statements more concisely (especially if we end up making your process class common code). It would simplify a lot of this code and make it more readable as well.

```
try
{
    GitHandler.Run(config, $"checkout -b {config.AssetsRepoBranch}")
    GitHandler.Run(config, $"add -A .")
    ...
}
catch GitProcessException ex
{
    throw GenerateInvokeException(ex.CommandResult);
}
```

Reply:
oh man I was legitimately thinking about this. The only reason I _haven't_ is because I don't know if I want individual exception handling or one catch all yet. This is due to the fact I haven't exercised this thing with many scenarios.

Reply:
I've made this update everywhere that invokes multiple commands in a row. I am not certain what the exception handling should look like, so I erred on the side of `beauty` :)

It looks far better than what I had before.